src/plugins/camera/camera_impl.cpp

```diff
@@ -760,25 +783,31 @@ void CameraImpl::process_camera_image_captured(const mavlink_message_t& message)
     mavlink_msg_camera_image_captured_decode(&message, &image_captured);
 
     {
-        std::lock_guard<std::mutex> lock(_capture_info.mutex);
+        Camera::CaptureInfo capture_info = {};
+        capture_info.position.latitude_deg = image_captured.lat / 1e7;
+        capture_info.position.longitude_deg = image_captured.lon / 1e7;
+        capture_info.position.absolute_altitude_m = image_captured.alt / 1e3f;
+        capture_info.position.relative_altitude_m = image_captured.relative_alt / 1e3f;
+        capture_info.time_utc_us = image_captured.time_utc;
+        capture_info.attitude_quaternion.w = image_captured.q[0];
+        capture_info.attitude_quaternion.x = image_captured.q[1];
+        capture_info.attitude_quaternion.y = image_captured.q[2];
+        capture_info.attitude_quaternion.z = image_captured.q[3];
+        capture_info.attitude_euler_angle =
+            to_euler_angle_from_quaternion(capture_info.attitude_quaternion);
+        capture_info.file_url = std::string(image_captured.file_url);
+        capture_info.is_success = (image_captured.capture_result == 1);
+        capture_info.index = image_captured.image_index;
+
+        if (capture_info.is_success) {
+            _status.photo_list.insert(std::make_pair(image_captured.image_index, capture_info));
+        }
 
-        if (_capture_info.callback) {
-            Camera::CaptureInfo capture_info = {};
-            capture_info.position.latitude_deg = image_captured.lat / 1e7;
-            capture_info.position.longitude_deg = image_captured.lon / 1e7;
-            capture_info.position.absolute_altitude_m = image_captured.alt / 1e3f;
-            capture_info.position.relative_altitude_m = image_captured.relative_alt / 1e3f;
-            capture_info.time_utc_us = image_captured.time_utc;
-            capture_info.attitude_quaternion.w = image_captured.q[0];
-            capture_info.attitude_quaternion.x = image_captured.q[1];
-            capture_info.attitude_quaternion.y = image_captured.q[2];
-            capture_info.attitude_quaternion.z = image_captured.q[3];
-            capture_info.attitude_euler_angle =
-                to_euler_angle_from_quaternion(capture_info.attitude_quaternion);
-            capture_info.file_url = std::string(image_captured.file_url);
-            capture_info.is_success = (image_captured.capture_result == 1);
-            capture_info.index = image_captured.image_index;
+        _captured_request_cv.notify_all();
 
+        std::lock_guard<std::mutex> lock(_capture_info.mutex);
+        if (_status.last_advertised_image_index < capture_info.index && _capture_info.callback) {
+            _status.last_advertised_image_index = capture_info.index;
```

Comment:
What's the relation between `_capture_info` and `_status`? and would we need both mutexes here?

Reply:
aha, yes you're right. Now I'm concerned about the risk of a deadlock if I use both :thinking:. We would need to be sure that we always acquire them in the same order, right?

Reply:
Actually, it should all be in `_capture_info`. I get the first part of your comment now :laughing: